src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -905,11 +905,23 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                 }
             }
 
-            if (typeParameter.HasUnmanagedTypeConstraint && (typeArgument.IsManagedType || !typeArgument.TypeSymbol.IsNonNullableValueType()))
+            if (typeParameter.HasUnmanagedTypeConstraint)
             {
-                // "The type '{2}' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as parameter '{1}' in the generic type or method '{0}'"
-                diagnosticsBuilder.Add(new TypeParameterDiagnosticInfo(typeParameter, new CSDiagnosticInfo(ErrorCode.ERR_UnmanagedConstraintNotSatisfied, containingSymbol.ConstructedFrom(), typeParameter, typeArgument.TypeSymbol)));
-                return false;
+                var managedKind = typeArgument.TypeSymbol.ManagedKind;
+                if (managedKind == ManagedKind.Managed || !typeArgument.TypeSymbol.IsNonNullableValueType())
+                {
+                    // "The type '{2}' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as parameter '{1}' in the generic type or method '{0}'"
+                    diagnosticsBuilder.Add(new TypeParameterDiagnosticInfo(typeParameter, new CSDiagnosticInfo(ErrorCode.ERR_UnmanagedConstraintNotSatisfied, containingSymbol.ConstructedFrom(), typeParameter, typeArgument.TypeSymbol)));
+                    return false;
+                }
+                else if (managedKind == ManagedKind.UnmanagedWithGenerics &&
+                    currentCompilation is CSharpCompilation csCompilation &&
```

Comment:
Can't this method just take a CSharpCompilation instead?

Reply:
We decided to use a cast instead because changing everywhere to use a CSharpCompilation looked like it would get out of hand.

---
In reply to: [246599179](https://github.com/dotnet/roslyn/pull/31602#discussion_r246599179) [](ancestors = 246599179)